### ISSUE TITLE
Enhancement: Evaluate different embedding models

### ISSUE THREAD

**Issue by rmusser01** [rmusser01/tldw#181]
Title.

**Comment by rmusser01** [comment]
Stella embedding model

https://huggingface.co/dunzhang/stella_en_1.5B_v5
https://huggingface.co/dunzhang/stella_en_400M_v5

**Comment by rmusser01** [comment]
Embeddings
https://www.youtube.com/watch?v=viZrOnJclY0
https://docs.llamaindex.ai/en/stable/examples/finetuning/embeddings/finetune_embedding/
https://modal.com/blog/fine-tuning-embeddings
https://huggingface.co/blog/train-sentence-transformers
https://www.philschmid.de/fine-tune-embedding-model-for-rag
https://arxiv.org/abs/2406.15241
https://arxiv.org/abs/2407.09252
https://aclanthology.org/W13-2322.pdf
https://towardsdatascience.com/how-to-reduce-embedding-size-and-increase-rag-retrieval-speed-7f903d3cecf7?gi=b5bcdb8dfecf
https://arxiv.org/abs/2304.01982

**Comment by rmusser01** [comment]
```
The basic gist is that we first use the LLM to generate better, more precise keywords that the RAG’s embedding model will be able to use to create an embedding vector closer to relevant matches. The LLM is run again with the more relevant info that the RAG found to hopefully generate a more accurate response.
```
https://medium.com/@djangoist/how-to-create-accurate-llm-responses-on-large-code-repositories-presenting-cgrag-a-new-feature-of-e77c0ffe432d
https://github.com/curvedinf/dir-assistant

**Comment by rmusser01** [comment]
https://github.com/amazon-science/hyperbolic-embeddings

**Comment by rmusser01** [comment]
https://www.reddit.com/r/LocalLLaMA/comments/1fhuktw/embedding_model_benchmark_code_with_autorag/

**Comment by rmusser01** [comment]
https://huggingface.co/papers/2409.10173

**Comment by rmusser01** [comment]
Evaluate swapping from Chroma to https://github.com/neuml/txtai
Also eval swapping to vec-sql

**Comment by rmusser01** [comment]
https://arxiv.org/pdf/2409.15700

**Comment by rmusser01** [comment]
https://huggingface.co/papers/2410.02525

**Comment by rmusser01** [comment]
https://github.com/tianyi-lab/MoE-Embedding
https://generativeai.pub/a-beginners-guide-to-fine-tuning-an-embedding-model-38bb4b4ae664
https://medium.com/@vici0549/it-is-crucial-to-properly-set-the-batch-size-when-using-sentence-transformers-for-embedding-models-3d41a3f8b649
https://blog.gopenai.com/fine-tuning-embeddings-for-specific-domains-a-comprehensive-guide-5e4298b42185

**Comment by rmusser01** [comment]
Fine tuning
https://docs.llamaindex.ai/en/stable/examples/finetuning/embeddings/finetune_embedding/

```
Last year my team worked on a fine tuned open source model, trained on US military doctrine and pubs ([workflow](https://www.reddit.com/r/LocalLLaMA/comments/15oome9/our_workflow_for_a_custom_questionanswering_app/) and [follow-up](https://www.reddit.com/r/LocalLLaMA/comments/1686ul6/some_lessons_learned_from_building_a_fine_tuned/) posts). Bottom line is that the fine tuned 7b model worked really well, especially on conceptual questions (like how maneuver and mission command interact): better than GPT-3.5 and about even with GPT-4 based on human ratings from military members.

Been itching to try fine tuning embeddings, and my team finally got a chance. We ran a series of experiments, but the big picture takeaway was that our first approach collapsed the embeddings space and made retrieval accuracy plummet, but a second approach using train+eval worked well and substantially improved retrieval.

We started with our model training data: a context+question column and answer column. We took the context chunk (500 tokens from a military publication) and the question generated from it, reversed their order and used them as the training data for the embeddings fine-tuning. So basically "When you see "What are the principles of air defense in urban areas?" then retrieve <some chunk about urban defense that has some sentences on air defense principles>.

We used Sentence Transformers and FSDP, because we had to shard the embedding model and data across multiple GPUs. To our distress however, each epoch of training made the model perform worse and worse, until at 5 epochs it was just random retrieval. Our intuition was that the model was overfitting and collapsing the embedding space until all documents were crammed next to each other. We used [WizMap](https://github.com/poloclub/wizmap/blob/main/LICENSE) to visualize embedded docs, and sure enough the base model showed clear clusters of docs, 2 epochs showed them kind of crammed closer, and at 5 epochs a giant blob with two camel humps.

We then switched to DDP from FSDP, which allows you to use an evaluator parameter during fine tuning, so we could use the eval data during training, not just post-hoc, something like:

    num_train_epochs=2,

    per_device_train_batch_size=32,

    per_device_eval_batch_size=32,

    During training, would train on a batch from the “TRAIN” dataset, and then evaluate on a batch from the “EVAL” dataet

    Use that train/eval comparison to inform the loss function

    Train for 2 or 5 epochs

    Post-training, ran our eval pipeline.

Success! Using BGE Small w. 384 dimensions, we went from:

    Base model top 20 accuracy of 54.4%.

    2 epochs fine-tuned model: Top 20 retrieval accuracy 70.8%.

    5 epochs fine-tuned model: Top 20 retrieval accuracy 73%.

We then tried Stella-400M 1024 dimensions:

    Base model top 20 accuracy of 62.9%.

    2 epochs fine-tuned model (train batch-size 4, gradient accumulation

    steps 20): Top 20 retrieval accuracy was 73.3%.

    3 epochs fine-tuned model (train batch-size 3, gradient accumulation

    steps 40): Top 20 retrieval accuracy was 72.4%

    Increased batch size (train batch size 8, grad accumulation steps 25) with 2

    epochs fine-tuning on 8 GPU clusters: Top 20 retrieval accuracy was 74.4%
```

https://www.reddit.com/r/LocalLLaMA/comments/15oome9/our_workflow_for_a_custom_questionanswering_app/
https://www.reddit.com/r/LocalLLaMA/comments/1686ul6/some_lessons_learned_from_building_a_fine_tuned/

**Comment by rmusser01** [comment]
```
This is a really tricky area of the field right now, because the current performance metrics we look for in embedding models are based on a set of ad-hoc metrics and random datasets that just so happened to be in vogue when the LLM sub-field started dominating the conversation a few years ago.

I’ve spent more hours the last two years than I can even describe on this, both personally and professionally, and here is how I currently think about this:

    The three axes to consider are concept obscurity, term volume, and top N precision.

    A model that performs well generally, aka on the MTEB leaderboard, is good at differentiating common concepts, when you have fewer terms to compare to one another, and when you’re comfortable with a “match” being in the top few results, not explicitly the first or second result.

    A more specialized model is the exact inverse, better on a set of highly specific, more obscure concepts, when you have a lot of them all at once, and when you need the top 1 or 2 matches to be “correct”.

Now, this gets even more fascinating, because there actually are real limits to how “good” a model can be on more common domains. And so, from my perspective, one simply considers the average term frequency of one’s domain relative to the dataset the model was trained on and can infer fitness from there.

Thus, models now are getting “better” at some more specialized domains because the datasets are larger and more inclusive of those sub-distributions. However, this scaling in “quality” does, from my testing, fall apart when the other two constraints come in.

So, long story short, use general models when you either have a “small” number of items to compare, OR are operating in a common domain, OR top N precision needs are loose. For most people, this is fine. For those of us in highly specialized domains where scale and precision are make or break factors, use a specialized model, up to and including creating your own.
```

**Comment by rmusser01** [comment]
https://emschwartz.me/binary-vector-embeddings-are-so-cool/
https://jkatz05.com/post/postgres/pgvector-scalar-binary-quantization/
https://www.mixedbread.ai/blog/binary-mrl

**Comment by rmusser01** [comment]
https://huggingface.co/nvidia/NV-Embed-v2/tree/main
